### DISCUSSION TITLE
Persist DB Connections

### DISCUSSION THREAD

**Discussion by cupcakearmy** [vercel/vercel#4776] [Help] [answered]
Hi everyone. I'm new to vercel (yersterday) and I was trying to check some things out.

My Problem: How do I persist DB Connection?
I followed [this official guide](https://vercel.com/guides/deploying-a-mongodb-powered-api-with-node-and-vercel) but it does not work for me using mongo. Every time the function gets called a new connection is built up. This slows down the response time to ~1s in dev and 500ms in prod. Which is basically unusable.

Any ideas/suggestions?
Thanks!!

Btw the code from the official guide looks like this:
```typescript
import { config } from 'dotenv'
const MongoClient = require('mongodb').MongoClient

config()
const { MONGO_URL, MONGO_DB, MONGO_USER, MONGO_PASSWORD } = process.env

let cachedDb = null

export async function getDB(uri: string = `mongodb+srv://${MONGO_USER}:${MONGO_PASSWORD}@${MONGO_URL}`) {
  if (cachedDb) {
    console.log('cached') // Never seen this in my logs :/
    return cachedDb
  }

  console.log('reconnecting')
  const client = await MongoClient.connect(uri, { useNewUrlParser: true, useUnifiedTopology: true })
  const db = await client.db(MONGO_DB)
  cachedDb = db

  return db
}
```

**Comment by davide-scalzo** [comment]
Hi @cupcakearmy , did you manage to solve this?

**Reply by cupcakearmy** [reply]
Unfortunately not.
But haven't tried it recently, maybe its fixed?

**Comment by cobrexus** [comment, marked as answer]
# From [Vercel's docs](https://vercel.com/docs/solutions/realtime):

> Realtime communication allows events to be streamed directly to your application by maintaining a single connection. WebSockets enable online experiences like collaborative document editing, multiplayer games, and interactive chatrooms.
>
> There are two main approaches to apply this model to stateless serverless functions.
>
> 1. [Serverless Functions](https://vercel.com/docs/serverless-functions/introduction) have maximum execution limits and should respond as quickly as possible. **They should not subscribe to data events.** Instead, we need a client that subscribes to data events and a serverless function that publishes new data. Consider using one of the providers below to implement this approach.
> 2. Rather than pushing data, you can fetch realtime data on-demand. For example, the Vercel dashboard delivers realtime updates using [SWR](https://swr.vercel.app).
>
> ## Providers
> These providers can help you integrate realtime communication into your Jamstack application using serverless functions.
> * [Pusher](https://github.com/vercel/vercel/tree/master/examples/vanilla-pusher-functions)
> * [PubNub](https://pubnub.com)
> * [Firebase Realtime Database](https://firebase.google.com/docs/database)
> * [TalkJS](https://talkjs.com)
> * [SendBird](https://sendbird.com)
> * [Supabase](https://github.com/vercel/next.js/tree/canary/examples/with-supabase-auth-realtime-db)

**Reply by cupcakearmy** [reply]
I was not listening to chages. I'm aware that lambas have a limited runtime in both cpu time and actual time. So cold starts of course will be slow, but also warm it was always reconnecting to the DB from scratch. One request after another. That said, might be different now

**Reply by cobrexus** [reply]
That's the way lambdas work, each function will open a new connection, and connections cannot be persisted. What were you trying to achieve?

**Reply by cupcakearmy** [reply]
Oh I see. Wast just testing around with a POC. Thanks for the clarification.
